lib/spack/spack/modules.py

```diff
@@ -259,3 +259,126 @@ def _write(self, m_file):
+
+    @property
+    def use_name(self):
+        pass
+
+    @staticmethod
+    def _compiler_module_directory(name, version):
+        return '{compiler_name}/{compiler_version}'.format(
+                compiler_name=name,
+                compiler_version=version
+            )
+
+    @staticmethod
+    def _mpi_module_directory(compiler_name, compiler_version, mpi_name, mpi_version):
+        return '{mpi_name}/{mpi_version}/{compiler_name}/{compiler_version}'.format(
+                mpi_name=mpi_name,
+                mpi_version=mpi_version,
+                compiler_name=compiler_name,
+                compiler_version=compiler_version
+            )
+
+    def _write(self, m_file):
+        # Header as in
+        # https://www.tacc.utexas.edu/research-development/tacc-projects/lmod/advanced-user-guide/more-about-writing-module-files
+        m_file.write("-- -*- lua -*-\n")
+        # Short description -> whatis()
+        if self.short_description:
+            m_file.write("whatis([[Name : {name}]])\n".format(name=self.spec.name))
+            m_file.write("whatis([[Version : {version}]])\n".format(version=self.spec.version))
+
+        # Long description -> help()
+        if self.long_description:
+            doc = re.sub(r'"', '\"', self.long_description)
+            m_file.write("help([[{documentation}]])\n".format(documentation=doc))
+
+        # Path alterations
+        for var, dirs in self.paths.items():
+            for directory in dirs:
+                m_file.write("prepend_path(\"{variable}\", \"{directory}\")\n".format(
+                    variable=var,
+                    directory=directory)
+                )
+        m_file.write("prepend_path(\"CMAKE_PREFIX_PATH\", \"{cmake_prefix}\")\n".format(cmake_prefix=self.spec.prefix))
+
+        # Add family protection
+        if self.family is not None:
+            m_file.write("family(\"{family}\")\n".format(family=self.family))
+
+        # Prepend path if family is 'compiler' or 'mpi'
+        if self.family is 'compiler':
+            hierarchy_name = self._compiler_module_directory(self.spec.name, self.spec.version)
+            fullname = join_path(LmodModule.path, hierarchy_name)
+            m_file.write("prepend_path(\"MODULEPATH\", \"{compiler_directory}\")".format(compiler_directory=fullname))
+        elif self.family is 'mpi':
+            s = self.spec
+            hierarchy_name = self._mpi_module_directory(s.compiler.name, s.compiler.version, s.name, s.version)
+            fullname = join_path(LmodModule.path, hierarchy_name)
+            m_file.write("prepend_path(\"MODULEPATH\", \"{compiler_directory}\")".format(compiler_directory=fullname))
+
+    def _use_system_compiler(self):
+        """
+        True if the spec uses the default compiler (which is assumed to be the one provided by the system)
+
+        :return: True or False
+        """
+        # FIXME: How can I check if a spec has been constructed using the system compiler?
```

Comment:
Yes, it's just the OS default compiler. To organize the modules hierarchically I have to select a default compiler (e.g. `/usr/bin/gcc`) which I will use:
* to build serial libraries that don't need compiler specific optimizations or don't benefit from them 
* to build generic tools, for instance for development purposes (e.g. cmake)
* to bootstrap other compilers if needed

The modules for those things will be placed in `<modules-root>/Core` and will be always visible to the user. On the other hand, modules that are built using a compiler provided in  `<modules-root>/Core` will be placed into their own directory and at most one directory will be visible to users at any time.